File: str/removeSpaces.java
//Q0: remove duplicate white space
//" I   love TV" -> "I love TV"
public String removeSpaces(String input) {
	if(input == null || input.length() == 0)
		return input
	char[] arr = input.toCharArray();
	int slow = 0;
	for(int fast = 0; fast < arr.length; fast++){
		//igore consecutive ' '
		//ignore first ' '
		if(arr[fast] == ' ' && (fast == 0 || arr[fast-1] == ' '))
			continue;
		else
			arr[slow++] = arr[fast];
	}
	//ignore last ' ', not catched in the for loop
	if(slow > 0 && arr[slow-1] == ' ')
		return new String(arr, 0, slow-1);
	return new String(arr, 0, slow);
}

File: str/removal/removeSpaces.py
# Q0: remove duplicate white space
#" I   love TV" -> "I love TV"
def removeSpaces(input):
	if input is None or not input:
		return input

	arr = list(input)
	slow = 0
	for fast in range(len(arr)):
		# ignore head or consecutive ' '
		if arr[fast] == ' ' and (fast == 0 or arr[fast-1] == ' '):
			continue
		else:
			arr[slow] = arr[fast]
			slow += 1
	# remove tail ' '
	# if slow > 0 and arr[slow-1] == ' ':
	# 	slow = slow - 1
	# return ''.join(arr[:slow]
	return ''.join(arr[:slow-1 if slow > 0 and arr[slow - 1] == ' ' else slow])
